File: app/src/main/java/com/ruideraj/secretelephant/send/SmsSender.java
package com.ruideraj.secretelephant.send;

import android.telephony.SmsManager;

public class SmsSender {

    private static volatile SmsSender INSTANCE;
    private static final Object sLock = new Object();

    private SmsManager mSmsManager;

    public static SmsSender getInstance(SmsManager smsManager) {
        if(INSTANCE == null) {
            synchronized(sLock) {
                if(INSTANCE == null) {
                    INSTANCE = new SmsSender(smsManager);
                }
            }
        }

        return INSTANCE;
    }

    public SmsSender(SmsManager smsManager) {
        mSmsManager = smsManager;
    }

    public void sendMessage(String recipient, String message) throws IllegalArgumentException {
        mSmsManager.sendTextMessage(recipient,
                null, message, null, null);
    }
}

File: app/src/main/java/com/ruideraj/secretelephant/send/SmsSender.kt
package com.ruideraj.secretelephant.send

import android.telephony.SmsManager
import com.ruideraj.secretelephant.BuildConfig
import com.ruideraj.secretelephant.PropertiesReader
import com.ruideraj.secretelephant.R
import com.ruideraj.secretelephant.contacts.Contact
import javax.inject.Inject

class SmsSender @Inject constructor(private val smsManager: SmsManager,
                                    private val propertiesReader: PropertiesReader): Sender {
    
    override suspend fun sendMessage(message: Message) {
        val contact = message.contact
        if (contact.type != Contact.Type.PHONE) throw IllegalArgumentException("Contact must be a phone number")

        val destination = if (BuildConfig.DEBUG) {
            propertiesReader.getProperty(R.string.test_send_key_phone)
        } else {
            ""  // TODO Add logic to use real phone number.
        }

        smsManager.sendTextMessage(destination, null, message.message, null, null)
    }
}